File: 백준/Gold/17298. 오큰수/오큰수.cc
#include <bits/stdc++.h>
using namespace std;

int result[1000001];

int main(void) {
    cin.tie(0)->sync_with_stdio(false);

    int n; cin >> n;
    stack<pair<int, int>> S;
    for (int i = 1; i <= n; i++) {
        int x; cin >> x;
        while (!S.empty() && S.top().second < x) {
            result[S.top().first] = x;
            S.pop();
        }
        S.push({i, x});
    }

    while (!S.empty()) {
        result[S.top().first] = -1;
        S.pop();
    }

    for (int i = 1; i <= n; i++) cout << result[i] << ' ';
}

File: 백준/Gold/17298. 오큰수/오큰수.kt
import java.util.*

/**
 * 내림차순 유지하다가 깨지는 순간 pop
 */

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val arr = StringTokenizer(readLine()).let { st -> IntArray(n) { st.nextToken().toInt() } }
    val stack = ArrayDeque<Pair<Int, Int>>()
    val result = IntArray(n)

    for ((i, x) in arr.withIndex()) {
        while (stack.isNotEmpty() && stack.peek().second < x) {
            val (idx, num) = stack.pop()
            result[idx] = x
        }
        stack.push(i to x)
    }

    while (stack.isNotEmpty()) {
        val (idx, num) = stack.pop()
        result[idx] = -1
    }

    print(result.joinToString(" "))
}